File: Algorithm/first_missing_positive.py
class Solution:
    def firstMissingPositive(self,A):
        index=0
        n=len(A)
        while index<n:
            if A[index]>0 and A[index]<=n and A[index]!=(index+1) and A[A[index]-1]!=A[index]:
                A[A[index]-1], A[index]=A[index],A[A[index]-1]
            else:
                index+=1
        for x in range(n):
            if A[x]!=x+1:
                return x+1
        return n+1

    def firstMissingPositive1 (self, A):
        n = len(A)
        for i in range(n):
            while  A[i]>0 and A[i]<= n  and A[i] != A[A[i]-1] :
                A[A[i]-1], A[i] = A[i], A[A[i]-1]
        for x in range(n):
            if A[x]!=x+1:
                return x+1
        return n+1


if __name__=='__main__':
    sol=Solution()
    A=[-1,4,2,1,9,10]
    A = [3,4,-1,1]
    B =  A[:]
    print sol.firstMissingPositive(A)
    print B
    print sol.firstMissingPositive1(B)

File: Algorithm/CPP/first_missing_positive.cpp
# include <vector>
# include <algorithm>
# include <string>
# include <stack>
# include <algorithm>
# include <deque>
# include <iostream>
# include <time.h>
using namespace std;
class Solution {
public:

	int firstMissingPostive(vector<int>& nums)
	{
		int i =0, n = nums.size();
		for(i=0; i<n; ++i)
		{
			while(nums[i]>0 && nums[i]<=n && nums[i] != nums[nums[i]-1])
			swap(nums[i], nums[nums[i]-1]); // loop until find the position,means the index plus one is the value
		}

		for(auto i=0; i<n; ++i)
		{
			if(nums[i] != i+1)
				return i+1;
		}
		return n+1;
	}

};

int main()
{
	Solution sol;
	srand((unsigned)time(NULL));

	vector<int>	 vec;
		for(int i=0; i<10;i++)
	{
		vec.push_back(rand()%11);         
	}
	//vec = {3, 4, -1, 1};
	for(auto v:vec) cout<<v<<" ";
	cout<<endl;
	cout<<sol.firstMissingPostive(vec)<<endl;

	system("pause");
	return 1;
}


